pkg/apis/apps/v1/defaults_test.go

```diff
@@ -298,20 +307,155 @@ func TestSetDefaultStatefulSet(t *testing.T) {
+						OnSetDeletion: appsv1.DeletePersistentVolumeClaimDeletePolicyType,
+					},
+				},
+			},
+			expected: &appsv1.StatefulSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: defaultLabels,
+				},
+				Spec: appsv1.StatefulSetSpec{
+					Replicas:            &defaultReplicas,
+					Template:            defaultTemplate,
+					PodManagementPolicy: appsv1.OrderedReadyPodManagement,
+					UpdateStrategy: appsv1.StatefulSetUpdateStrategy{
+						Type: appsv1.RollingUpdateStatefulSetStrategyType,
+						RollingUpdate: &appsv1.RollingUpdateStatefulSetStrategy{
+							Partition: &defaultPartition,
+						},
+					},
+					PersistentVolumeClaimPolicy: &appsv1.StatefulSetPersistentVolumeClaimPolicy{
+						OnSetDeletion: appsv1.DeletePersistentVolumeClaimDeletePolicyType,
+						OnScaleDown:   appsv1.RetainPersistentVolumeClaimDeletePolicyType,
+					},
+					RevisionHistoryLimit: utilpointer.Int32Ptr(10),
+				},
+			},
+			enablePVCDeletionPolicy: true,
+		},
+		{
+			name: "PVC delete policy disabled, with policy specified",
+			original: &appsv1.StatefulSet{
+				Spec: appsv1.StatefulSetSpec{
+					Template: defaultTemplate,
+					PersistentVolumeClaimPolicy: &appsv1.StatefulSetPersistentVolumeClaimPolicy{
+						OnScaleDown: appsv1.DeletePersistentVolumeClaimDeletePolicyType,
+					},
+				},
+			},
+			expected: &appsv1.StatefulSet{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: defaultLabels,
+				},
+				Spec: appsv1.StatefulSetSpec{
+					Replicas:            &defaultReplicas,
+					Template:            defaultTemplate,
+					PodManagementPolicy: appsv1.OrderedReadyPodManagement,
+					UpdateStrategy: appsv1.StatefulSetUpdateStrategy{
+						Type: appsv1.RollingUpdateStatefulSetStrategyType,
+						RollingUpdate: &appsv1.RollingUpdateStatefulSetStrategy{
+							Partition: &defaultPartition,
+						},
+					},
+					PersistentVolumeClaimPolicy: &appsv1.StatefulSetPersistentVolumeClaimPolicy{
+						OnScaleDown: appsv1.DeletePersistentVolumeClaimDeletePolicyType,
+					},
+					RevisionHistoryLimit: utilpointer.Int32Ptr(10),
+				},
+			},
+			enablePVCDeletionPolicy: false,
+		},
 	}
 
-	for i, test := range tests {
-		original := test.original
-		expected := test.expected
-		obj2 := roundTrip(t, runtime.Object(original))
-		got, ok := obj2.(*appsv1.StatefulSet)
-		if !ok {
-			t.Errorf("(%d) unexpected object: %v", i, got)
-			t.FailNow()
-		}
-		if !apiequality.Semantic.DeepEqual(got.Spec, expected.Spec) {
-			t.Errorf("(%d) got different than expected\ngot:\n\t%+v\nexpected:\n\t%+v", i, got.Spec, expected.Spec)
-		}
+	for _, test := range tests {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			if test.enablePVCDeletionPolicy {
+				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StatefulSetAutoDeletePVC, true)()
+			} else {
+				defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StatefulSetAutoDeletePVC, false)()
+			}
```

Comment:
```golang
defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.StatefulSetAutoDeletePVC, test.enablePVCDeletionPolicy)()
```
:)

Reply:
gosh I've been staring at this code too long :-)